File: Difficulty: Medium/LCA in Binary Tree/lca-in-binary-tree.cpp
/* A binary tree node

struct Node
{
    int data;
    struct Node* left;
    struct Node* right;

    Node(int x){
        data = x;
        left = right = NULL;
    }
};
 */

class Solution {
  public:
    // Function to return the lowest common ancestor in a Binary Tree.
    Node* dfs(Node* root,int n1, int n2) {
        if(root==nullptr) return nullptr;
        if(root->data == n1 || root->data==n2) return root;
        Node* left = dfs(root->left,n1,n2);
        Node* right = dfs(root->right,n1,n2);
        
        if(left!=nullptr && right!=nullptr) return root;
        if(left==nullptr) return right;
        return left;
        
    }
    Node* lca(Node* root, int n1, int n2) {
        // Your code here
        return dfs(root,n1,n2);
        
    }
};

File: Difficulty: Medium/LCA in Binary Tree/lca-in-binary-tree.java
/* Structure of binary tree node
class Node {
    int data;
    Node left;
    Node right;

    Node(int val) {
        this.data = val;
        left = right = null;
    }
}
}*/

class Solution {
    
    private Node lcaSolve(Node root, int a, int b) {
        if(root== null) return root;
        
        if(root.data==a || root.data==b) return root;
        
        Node left = lcaSolve(root.left, a, b);
        Node right = lcaSolve(root.right,a,b);
        
        if((left!=null && (left.data==a || left.data==b)) &&
            (right!=null && (right.data ==a || right.data==b))) {
                return root;
        }
        
        if(left==null) return right;
        return left;
    }
    Node lca(Node root, int n1, int n2) {
        // code here
        Node ans = lcaSolve(root,n1,n2);
        return ans;
    }
}